api/STRUCTURE.md

```diff
@@ -0,0 +1,84 @@
+
+TODO: do we have a preference of which of those to do?
+
+## DB Models
+
+DB Models represent a single row of a single table in the database, possibly with extender tables attached. We denote DB models by prefixing `Db` to the name of the database table. DB models heavily use `javax.persistence` annotations to describe where in the database table each part of the model lives.
+
+In order to automatically gather data from extender tables, we use the `@JoinColumn`, `@OneToOne`, `@OneToMany`, `@ManyToOne`, and `@MappedBy` annotations. Documentation and blogs on the usage of these annotations can be found in these links:
+
+- https://www.baeldung.com/jpa-join-column
+- https://www.baeldung.com/hibernate-one-to-many
+- https://www.baeldung.com/jpa-joincolumn-vs-mappedby
+- https://docs.jboss.org/hibernate/jpa/2.1/api/javax/persistence/JoinColumn.html (contains See Also links to the rest of these annotations)
+
+Examples of the use of these annotations can be found in the following places:
+
+- [DbUser](https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/db/model/DbUser.java)
+- [DbAddress](https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/db/model/DbAddress.java)
+- [DbCohort](https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/db/model/DbCohort.java)
+
+Much of the structure of DB models can be autogenerated in IntelliJ by defining the class members and then doing Generate -> Getter and Setter and Generate -> `equals()` and `hashCode()`. This will not automatically add the required annotations - those must be added manually.
+
+## Services and ServiceImpls
+
+Services are interfaces that declare methods of business logic. ServiceImpls are classes implementing their correspondingly named Service and the business logic methods declared therein.
+
+Practically, this allows one ServiceImpl to implement multiple interfaces. This is used in several `UserServiceImpl` and others to mixin [GaugeDataCollector](https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/monitoring/GaugeDataCollector.java), which needs a different implementation per-service.
+
+Services / ServiceImpls are generally named after a feature or functional area, like 'UserService', 'CohortService', or 'MonitoringService'.
+
+ServiceImpls can call into other services, but we try to keep the number of dependencies down. We try to avoid having ServiceImpls call into a large number of DAOs, especially DAOs not directly related to the functional area for which the ServiceImpl is responsible. For instance, the `UserServiceImpl` calls into the `UserDao`, the `AdminActionHistoryDao`
```

Comment:
The last sentence looks incomplete.  Also, is this intended as a positive example because it only talks to two DAOs or a negative one because we'd prefer it to talk to only one?

Reply:
Positive example because everything it talks to is directly related to the purpose of the service. Something massive like `WorkspaceServiceImpl` should probably 1) get broken up and 2) stop talking to the `UserDao` directly.

Reply:
Ideally, each DAO would roll up to only one Service.

Reply:
No, you could certainly imagine a service that needs more than one table (dao), or a table that's useful to more than one service. For example, the CohortDao is injected into a couple of services.

But yes, it makes lots of sense to break up some of these mega-services, even if they share DAOs.